gcm_toolkit/utils/clouds.py

```diff
@@ -0,0 +1,364 @@
+                            ref_index[wav, 1] = 0
+                        else:
+                            ref_index[wav, 1] = np.exp(np.log(klo) + fac * np.log(khi/klo))
+                        break
+                else:
+                    # if wavelength is outside of range, extrapolate
+                    if header == '.True.':
+                        # conducting interpolation, log-log interpolation
+                        # find 70% wavelength
+                        cur_wave = float(data[0].split()[0])  # current wavelength
+                        max_wave = float(data[-1].split()[0])  # maximum wavelength
+                        c_nr = 0
+                        for dnr, _ in enumerate(data):
+                            cur_wave = float(data[dnr].split()[0])  # current wavelength
+                            if cur_wave < 0.7 * max_wave:
+                                break
+                            c_nr += 1
+                        nlo = float(data[c_nr].split()[1])  # lower n value
+                        nhi = float(data[-1].split()[1])  # higher n value
+                        klo = float(data[c_nr].split()[2])  # lower k value
+                        khi = float(data[-1].split()[2])  # higher k value
+
+                        # calculate interpolation
+                        fac = np.log(wave/max_wave)/np.log(cur_wave/max_wave)
+                        ref_index[wav, 0] = np.exp(np.log(nhi) + fac * np.log(nlo/nhi))
+                        if klo <= 0 or khi < 0:
+                            ref_index[wav, 1] = 0
+                        else:
+                            ref_index[wav, 1] = np.exp(np.log(khi) + fac * np.log(klo/khi))
+                    else:
+                        # non-conducting interpolation, linear decreasing k, constant n
+                        ref_index[wav, 0] = float(data[-1].split()[1])
+                        ref_index[wav, 1] = float(data[-1].split()[2])*\
+                                            float(data[-1].split()[0])/wave
+
+            # stop searching the files
+            break
+    else:
+        # if no data was found, throw error
+        wrt.write_status(
+            "WARN",
+            'The following cloud particle bulk species is currently ' +
+            'not supported or misspelled: ' + species + '. Vacuum values ' +
+            'are used instead'
+            )
+
+    # return refractive indices
+    return ref_index
+
+
+def patch_cloud_mix_opa(self, clouds=None):
+    """
+    This function is used to update petitRADTRANS to include clouds.
+    """
+    if clouds is not None:
+        lsk = self.line_struc_kappas
+        self.line_struc_kappas = np.zeros((lsk.shape[0], lsk.shape[1],
+                                           lsk.shape[2]+1, lsk.shape[3]))
+        self.line_struc_kappas[:, :, :-1, :] = lsk
+        for j in range(len(self.line_struc_kappas[:, 0, 0, 0])):
+            self.line_struc_kappas[j, :, -1, :] = clouds[0]
```

Comment:
such a clever function! I assume you tested it?

You could get rid of the loop using
```suggestion
self.line_struc_kappas[:, :, -1, :] = np.ones_like(self.line_struc_kappas[:, :, -1, :])*clouds[0][np.newaxis,:,:]
```
But yours is oc fine as well!

Reply:
If tested it on HATS-6b and the results look reasonable. I also run tests on submodules which should be correct. I did not yet benchmarked it against other cloud codes which might be difficult.